### PR TITLE
Topic2 Exersice 1

### PR DESCRIPTION
**Pull request by marianoorio**

### CODE REVIEW

File: Topic3-Services/src/main/java/com/bootcamp/topic3/ServiceShoppingCart/Item.java [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+ * 
+ *
+ */
+public class Item {
+	private String itemId;
+	private String itemDescription;
+	private double itemPrice;
+	private int itemStockQuantity;
+	
+	public Item(String itemId, String itemDescription, double itemPrice, int itemStockQuantity){
```
Comment (sebastian-garofalo) [inline Topic3-Services/src/main/java/com/bootcamp/topic3/ServiceShoppingCart/Item.java]: For prices float is enough unless you're selling planets or something like that, check this article:
http://cs-fundamentals.com/java-programming/java-primitive-data-types.php

File: Topic3-Services/src/main/java/com/bootcamp/topic3/ServiceShoppingCart/Item.java [hunk 2]
```diff
@@ -0,0 +1,49 @@
+package com.bootcamp.topic3.ServiceShoppingCart;
+
+/**
```
Comment (sebastian-garofalo) [inline Topic3-Services/src/main/java/com/bootcamp/topic3/ServiceShoppingCart/Item.java]: remember to add a meaningful class description in the javadoc

File: Topic3-Services/src/main/java/com/bootcamp/topic3/ServiceShoppingCart/ServiceShoppingCartImplementation.java [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+		for (String key : cartItems.keySet()){
+			itemList.add(cartItems.get(key).getItem());
+		}
+		return itemList;
+	}
+
+	@Override
+	public int getItemQuantity(Item item) {
+		if (item != null && cartItems.containsKey(item.getItemId())){
+			return cartItems.get(item.getItemId()).getQuantity();
```
Comment (sebastian-garofalo) [inline Topic3-Services/src/main/java/com/bootcamp/topic3/ServiceShoppingCart/ServiceShoppingCartImplementation.java]: Its a good practice to only have one return in your methods.
That helps static code analysis tools to work better and improve readability for the human

File: Topic3-Services/src/main/java/com/bootcamp/topic3/ServiceUsers/ServiceUsers.java [hunk 4]
```diff
@@ -0,0 +1,38 @@
+package com.bootcamp.topic3.ServiceUsers;
+
+/**
+ * 
+ * Interface for users services
+ * CRUD Operations
+ *
+ */
+
```
Comment (sebastian-garofalo) [inline Topic3-Services/src/main/java/com/bootcamp/topic3/ServiceUsers/ServiceUsers.java]: There shouldn't be an space between the javadoc and the method declaration

File: Topic3-Services/src/main/java/com/bootcamp/topic3/ServiceUsers/ServiceUsersFactory.java [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+ */
+public class ServiceUsersFactory {
+	
+	private ServiceUsersFactory(){};
+	
+	public static ServiceUsers getShoppingCartService(ServiceUsersTypes type) {
+		if (ServiceUsersTypes.LOCAL_SERVICE.equals(type)) {
+			return new ServiceUsersImplementation();
+		}
+		return null;
```
Comment (sebastian-garofalo) [inline Topic3-Services/src/main/java/com/bootcamp/topic3/ServiceUsers/ServiceUsersFactory.java]: In this case maybe is better to throw an illegal parameter exception than to return null.

File: Topic3-Services/src/main/java/com/bootcamp/topic3/ServiceUsers/ServiceUsersImplementation.java [hunk 6]
```diff
@@ -0,0 +1,54 @@
+package com.bootcamp.topic3.ServiceUsers;
+
+import java.util.Hashtable;
+import java.util.Map;
+
+
+public class ServiceUsersImplementation implements ServiceUsers {
```
Comment (sebastian-garofalo) [inline Topic3-Services/src/main/java/com/bootcamp/topic3/ServiceUsers/ServiceUsersImplementation.java]: missing javadoc

File: topic2/src/main/java/com/bootcamp/topic2/exersice1/RecentFiles.java [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+	List<String> pathList;
+	
+	
+	/**
+	 * Constructor
+	 */
+	public RecentFiles(){
+		//ArrayList is inefficient to add always on the first position?? 
+		pathList = new ArrayList<String>();
+		maxSize = 15;
```
Comment (sebastian-garofalo) [inline topic2/src/main/java/com/bootcamp/topic2/exersice1/RecentFiles.java]: Please read about "magic numbers".

Always try to store values in an specific constant.

File: topic2/src/main/java/com/bootcamp/topic2/exersice2/Post.java [hunk 8]
```diff
@@ -0,0 +1,52 @@
+package com.bootcamp.topic2.exersice2;
+
+public class Post {
```
Comment (sebastian-garofalo) [inline topic2/src/main/java/com/bootcamp/topic2/exersice2/Post.java]: A lot of class missing the javadocs

File: topic2/src/test/java/com/bootcamp/topic2/exersice1Test/FileManagerTest.java [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+	
+	FileManager fileManager = FileManager.getFileManager();
+	
+	@Before
+	public void setUp() throws Exception {
+		//spy the class because it's applying a singleton pattern and can't inject mock
+		fileManager = Mockito.spy(fileManager);
+		//Mockito.when(fileManager.getBufferedReader(Mockito.any(File.class))).thenReturn(null);
+		//Alternative when.then way
+		Mockito.doReturn(null).when(fileManager).getBufferedReader(Mockito.any(File.class));
```
Comment (marcoStella) [inline topic2/src/test/java/com/bootcamp/topic2/exersice1Test/FileManagerTest.java]: Use as a general rule when you add a "when" statement for mocking behavior to have a "verify" to validate that what you are mocking is done. Here if you add it in the "Before" you are assuming that that behavior will be used in all the tests. Due that in all the tests of the class we should see something like this (or in the After):

```
Mockito.verify(fileManager, Mockito.times(the amount of times the mocked behavior is used)).getBufferedReader(Mockito.any(File.class))
```
If that behavior only applies to some of the tests, remove it from the Before and add it in the tests that it is used

File: Topic3-Services/src/main/java/com/bootcamp/topic3/ServiceShoppingCart/Item.java (hunk 2 above)
Comment (marianoorio) [inline Topic3-Services/src/main/java/com/bootcamp/topic3/ServiceShoppingCart/Item.java]: Added the missing javadocs.

File: Topic3-Services/src/main/java/com/bootcamp/topic3/ServiceShoppingCart/Item.java (hunk 1 above)
Comment (marianoorio) [inline Topic3-Services/src/main/java/com/bootcamp/topic3/ServiceShoppingCart/Item.java]: Planets sales could be the next boom market and we will need be able to work with it. 

The truths it's the compiler gave me an error when I tried to put literals floats on the constructors and I solved it with doubles. I didn't know I have to put f on the end of a literal float

File: Topic3-Services/src/main/java/com/bootcamp/topic3/ServiceShoppingCart/ServiceShoppingCartImplementation.java (hunk 3 above)
Comment (marianoorio) [inline Topic3-Services/src/main/java/com/bootcamp/topic3/ServiceShoppingCart/ServiceShoppingCartImplementation.java]: I fixed it and it makes me refactor the if statement and now it's more clear.

File: Topic3-Services/src/main/java/com/bootcamp/topic3/ServiceUsers/ServiceUsers.java (hunk 4 above)
Comment (marianoorio) [inline Topic3-Services/src/main/java/com/bootcamp/topic3/ServiceUsers/ServiceUsers.java]: Fixed

File: Topic3-Services/src/main/java/com/bootcamp/topic3/ServiceUsers/ServiceUsersFactory.java (hunk 5 above)
Comment (marianoorio) [inline Topic3-Services/src/main/java/com/bootcamp/topic3/ServiceUsers/ServiceUsersFactory.java]: Added an exception class for this. Always when could get an object or service with http request it's better throw an exception than return null?

File: Topic3-Services/src/main/java/com/bootcamp/topic3/ServiceUsers/ServiceUsersImplementation.java (hunk 6 above)
Comment (marianoorio) [inline Topic3-Services/src/main/java/com/bootcamp/topic3/ServiceUsers/ServiceUsersImplementation.java]: Fixed

File: topic2/src/main/java/com/bootcamp/topic2/exersice1/RecentFiles.java (hunk 7 above)
Comment (marianoorio) [inline topic2/src/main/java/com/bootcamp/topic2/exersice1/RecentFiles.java]: I focused on if the size must be constant or not and I forgot to set the magic '15' with a default constant. Anyway, I readed about magic numbers to refresh the idea

Comment (marianoorio) [inline topic2/src/main/java/com/bootcamp/topic2/exersice1/RecentFiles.java]: read*

File: topic2/src/main/java/com/bootcamp/topic2/exersice2/Post.java (hunk 8 above)
Comment (marianoorio) [inline topic2/src/main/java/com/bootcamp/topic2/exersice2/Post.java]: Fixed. Thanks for all comments

File: Topic3-Services/src/test/java/com/bootcamp/topic3/ServiceShoppingCart/ServiceShoppingCartTest.java [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+	public void whenCreatesACartItIsEmpty() {
+		shopCart = new ServiceShoppingCartImplementation();
+		Assert.assertTrue(shopCart.isEmpty());
+	}
+	
+	@Test
+	public void whenAddsAnItemToCartItIsAdded() {
+		Item item = new Item("Item0","test item", 00, 12 );
+		shopCart.addToCart(item,3);
+		Assert.assertEquals(item, shopCart.getAllItems().get(0));
```
Comment (marianoorio) [inline Topic3-Services/src/test/java/com/bootcamp/topic3/ServiceShoppingCart/ServiceShoppingCartTest.java]: Ok, with the eclipse's auto override methods it's enough?

File: Topic3-Services/src/test/java/com/bootcamp/topic3/ServiceUsers/ServiceUsersTest.java [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+	@Test
+	public final void whenReadsAnUnexistentUserItReturnNull() {
+		Assert.assertNull(users.readUser("user0"));
+	}
+	
+	@Test
+	public final void whenReadsAnExistentUserItReturnTheUser() {
+		User user0 = new User("user0", "passwordUser0","user","user@test.com");
+		users.createUser(user0);
+		Assert.assertEquals(user0, users.readUser(user0.getUserName()));
```
Comment (marianoorio) [inline Topic3-Services/src/test/java/com/bootcamp/topic3/ServiceUsers/ServiceUsersTest.java]: Fixed

File: topic2/src/test/java/com/bootcamp/topic2/exersice1Test/FileManagerTest.java (hunk 9 above)
Comment (marianoorio) [inline topic2/src/test/java/com/bootcamp/topic2/exersice1Test/FileManagerTest.java]: I carried the error from tests FileManager and RecentsFiles as one and in each tests I opened files, so I put that config on @before. Thanks for all comments

File: Topic3-Services/src/main/java/com/bootcamp/topic3/ServiceShoppingCart/Item.java [hunk 12]
```diff
@@ -1,20 +1,19 @@
...
  */
 public class Item {
 	private String itemId;
 	private String itemDescription;
-	private double itemPrice;
+	private float itemPrice;
 	private int itemStockQuantity;
 	
-	public Item(String itemId, String itemDescription, double itemPrice, int itemStockQuantity){
+	public Item(String itemId, String itemDescription, float d, int itemStockQuantity){
```
Comment (marianoorio) [inline Topic3-Services/src/main/java/com/bootcamp/topic3/ServiceShoppingCart/Item.java]: eclipse modified my parameters for work correctly when I did changes on tests first and I miss it

File: Topic3-Services/src/main/java/com/bootcamp/topic3/ServiceShoppingCart/ServiceNotFoundException.java [hunk 13]
```diff
@@ -0,0 +1,18 @@
...
+ * 
+ * Exception thrown when a service are not found
+ *
+ */
+public class ServiceNotFoundException extends Exception{
+	
+	/**
+	 * Serialized because the inheritance of a serialized class
+	 */
+	private static final long serialVersionUID = 1L;
```
Comment (marianoorio) [inline Topic3-Services/src/main/java/com/bootcamp/topic3/ServiceShoppingCart/ServiceNotFoundException.java]: I used the autoGenerated from eclipse to that. As I told Mauricio when he called me, I have to read more about serialization after complete the bootcamp to understand more of it.

**Pull request merged**